### PR TITLE
Move user-defined -idirafter to the end of args

### PR DESCRIPTION
**Pull request by tanshihaj**
###### Description of changes

As discussed in https://github.com/NixOS/nixpkgs/issues/190005 we have an issue with `-idirafter` arg in cc-wrapper. In this PR we add logic that moves all `-idirafter` to the end of arg list.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/cc-wrapper/cc-wrapper.sh [hunk 1]
```diff
@@ -135,6 +135,23 @@ source @out@/nix-support/add-hardening.sh
 extraAfter=($NIX_CFLAGS_COMPILE_@suffixSalt@)
 extraBefore=(${hardeningCFlags[@]+"${hardeningCFlags[@]}"} $NIX_CFLAGS_COMPILE_BEFORE_@suffixSalt@)
 
+# Move all user-defined -idirafter from params to the extraAfter, so -idirafter with libs include
+# paths will be defined before user-defined -idirafter
+kept=()
+n=0
+while (( $n < ${#params[@]} )); do
+    if [[ ${params[n]} =~ ^\-idirafter$ ]]; then
+        extraAfter+=("${params[n]}" "${params[n+1]}")
```
Comment (tanshihaj) [inline pkgs/build-support/cc-wrapper/cc-wrapper.sh:144]: I forget to check if `${params[n+1]}` exists - will fix it.

**Comment by tanshihaj** [comment]
@Ericson2314 Do I need some fixes here before merge?

**Comment by tanshihaj** [comment]
@veprbl Can you help me with this PR? How can I help with review?

**Comment by veprbl** [comment]
I looked at the code, and it looks fine, although it may be combined with a loop above. I have some questions regarding the intention of this.

This is supposed to allow a user to break our "soft compiler sandbox" using -idirafter, right? In the original issue you've mentioned how libSystem libc is in your way and mentioned u-boot. It makes me wonder if the `stdenv` of your choice is the problem. Why are you not using, say, `pkgsCross.gnu64.stdenv`? If using darwin `stdenv` is a must, you can always mangle `NIX_CFLGAGS_COMPILE` in your shell.nix to exclude the paths you don't lile.

**Comment by tanshihaj** [comment]
> This is supposed to allow a user to break our "soft compiler sandbox" using -idirafter, right?

No, it just copy behaviour of default clang/gcc. Details here: https://github.com/NixOS/nixpkgs/issues/190005, TLDR: user-defined `-idirafter` should go _after_ libc's include paths.

**Comment by veprbl** [comment]
Can the same be achieved by reversing order of concatenated items in
https://github.com/NixOS/nixpkgs/blob/0446732359536566de33d8893bd4c682174e31e2/pkgs/build-support/cc-wrapper/add-flags.sh#L51
?

**Comment by tanshihaj** [comment]
No, unfortunately not. We need to move all `-idirafter` flags from `params` to end of `extraAfter`. Changing order in `NIX_CFLAGS_COMPILE_@suffixSalt@` will just move them inside `extraAfter` environment variable.

But I just realized that maybe it would be good idea to move `-idirafter` with libc include paths to separate variable and put it to `extraBefore`, that will make `while (( $n < $N )); do` unnecessary and source code will look better.

**Comment by tanshihaj** [comment]
I've tried to implement solution I described above and now I cannot compile `glibc`. We cannot just pull `libc-cflags` (which contains `-idirafter` with libc include dirs) and put it into `extraBefore`: `crt1-cflags`, `libc-cflags`, `cc-cflags` must remain in specific order:
https://github.com/NixOS/nixpkgs/blob/0446732359536566de33d8893bd4c682174e31e2/pkgs/build-support/cc-wrapper/add-flags.sh#L36-L41

So the only solution is to collect all user-defined `-idirafter` and put them into end of `extraAfter`, which is already implemented in PR.

**Review by veprbl — APPROVED** [review]
This addresses an edge case, but the proposed solution seems sufficiently local to not be a huge burden.

**Comment by tanshihaj** [comment]
Yes, I agree that solution in pr is a bit ugly, will try to find another way to do it.

**Comment by tanshihaj** [comment]
Ok, I give up - unfortunately I do not have enough knowledge (or time to get this knowledge) to fix both https://github.com/NixOS/nixpkgs/issues/190005 and https://github.com/NixOS/nixpkgs/issues/158042 in one solution. Closing PR.

**Pull request closed without merging**